src/java.base/share/classes/sun/security/provider/Dilithium.java

```diff
@@ -0,0 +1,1432 @@
+                level = 3;
+                tau = 49;
+                lambda = 192;
+                gamma1 = 1 << 19;
+                gamma1Bits = 19;
+                gamma2 = (dilithium_q - 1) / 32;
+                dilithium_k = 6;
+                dilithium_l = 5;
+                eta = 4;
+                beta = 196;
+                omega = 55;
+                privateKeyLength = 0;
+                publicKeyLength = 1952;
+                signatureLength = 3293;
+                s1PackedLength = 640;
+                s2PackedLength = 768;
+                s1s2CoeffSize = 4;
+                wCoeffSize = 4;
+                break;
+            case 5:
+                level = 4;
+                tau = 60;
+                lambda = 256;
+                gamma1 = 1 << 19;
+                gamma1Bits = 19;
+                gamma2 = (dilithium_q - 1) / 32;
+                dilithium_k = 8;
+                dilithium_l = 7;
+                eta = 2;
+                beta = 120;
+                omega = 75;
+                privateKeyLength = 0;
+                publicKeyLength = 2592;
+                signatureLength = 4595;
+                s1PackedLength = 672;
+                s2PackedLength = 768;
+                s1s2CoeffSize = 3;
+                wCoeffSize = 4;
+                break;
+            default:
+                throw new IllegalArgumentException("Wrong security level");
+        }
+    }
+
+    public record DilithiumPrivateKey(byte[] rho, byte[] k, byte[] tr, int[][] s1, int[][] s2, int[][] t0) {
+    }
+
+    public record DilithiumPublicKey(byte[] rho, int[][] t1) {
+    }
+
+    public record DilithiumKeyPair(DilithiumPrivateKey privateKey, DilithiumPublicKey publicKey) {
+    }
+
+    public record DilithiumSignature(byte[] commitmentHash, int[][] response, boolean[][] hint) {
+    }
+
+    public DilithiumKeyPair generateKeyPair(byte[] randomBytes) {
+        //Initialize hash functions
+        var dilithiumH = new SHAKE256(0);
+        var dilithiumCRH = new SHAKE256(dilithiumTrLength);
+
+        //Expand seed
+        dilithiumH.update(randomBytes);
+        dilithiumH.update((byte)dilithium_k);
+        dilithiumH.update((byte)dilithium_l);
+        byte[] rho = dilithiumH.squeeze(dilithiumASeedLength);
+        byte[] rhoPrime = dilithiumH.squeeze(dilithiumS1S2SeedLength);
+        byte[] k = dilithiumH.squeeze(dilithiumKLength);
+        dilithiumH.reset();
+
+        //Sample A
+        int[][][] keygenA = generateA(rho); //A is in NTT domain
+
+        //Sample S1 and S2
+        int[][] s1 = new int[dilithium_l][dilithium_n];
+        int[][] s2 = new int[dilithium_k][dilithium_n];
+        sampleS1S2(s1, s2, dilithiumH, rhoPrime); //dilithiumH is reset before being used in sampleS1S2
+
+        //Compute t and tr
+        mlDsaVectorNtt(s1); //s1 now in NTT domain
+        int[][] As1 = matrixVectorPointwiseMultiply(keygenA, s1);
+        mlDsaVectorInverseNtt(s1); //take s1 out of NTT domain
```

Comment:
`s1` is back to its sampled values. Is it worth creating a `mlDsaVectorNtt` that does not modify in-place? Then we don't need to call NTTinv.